### PR TITLE
Fetching build secrets and putting them into the environment

### PR DESCRIPTION
**Pull request by jer**
- Also removing SD_TOKEN from the environment to prevent accidental
  leakage
- Refactored the environment code a little bit while I was at it

Fixes https://github.com/screwdriver-cd/launcher/issues/72

### CODE REVIEW

File: screwdriver/screwdriver.go [hunk 1]
```diff
@@ -435,3 +445,24 @@ func (a api) UpdateStepStop(buildID, stepName string, exitCode int) error {
 
 	return nil
 }
+
+func (a api) SecretsForBuild(build Build) (Secrets, error) {
+	u, err := a.makeURL(fmt.Sprintf("builds/%s/secrets", build.ID))
```
Comment (minz1027) [inline screwdriver/screwdriver.go]: Just to clarify, does this mean we need an API endpoint like `GET build/{id}/secrets`?

Comment (jer) [inline screwdriver/screwdriver.go]: Yes, that's the one that @stjohnjohnson is working on

**Comment by petey** [comment]
👍  Looks good to me.

File: executor/executor.go [hunk 2]
```diff
@@ -58,12 +57,8 @@ func doRun(cmd screwdriver.CommandDef, emitter screwdriver.Emitter, env []string
 }
 
 // Run executes a slice of CommandDefs
-func Run(path string, emitter screwdriver.Emitter, job screwdriver.JobDef, api screwdriver.API, buildID string) error {
+func Run(path string, env []string, emitter screwdriver.Emitter, job screwdriver.JobDef, api screwdriver.API, buildID string) error {
```
Comment (d2lam) [inline executor/executor.go]: where do we use `env` in this function?

Comment (stjohnjohnson) [inline executor/executor.go]: Line 68

File: screwdriver/screwdriver.go (hunk 1 above)
Comment (d2lam) [inline screwdriver/screwdriver.go]: Is this a DNM then since the endpoint is not there yet?

File: launch_test.go [hunk 3]
```diff
@@ -965,22 +940,78 @@ func TestSetEnv(t *testing.T) {
...
+
+	testBuild := FakeBuild{ID: TestBuildID, JobID: TestJobID}
+	api.secretsForBuild = func(build screwdriver.Build) (screwdriver.Secrets, error) {
+		if !reflect.DeepEqual(build.ID, testBuild.ID) {
+			t.Errorf("build.ID = %v, want %v", build.ID, testBuild.ID)
+		}
 
-		if !reflect.DeepEqual(err, test.err) {
-			t.Errorf("Error = %v, want %v", err, test.err)
+		testSecrets := screwdriver.Secrets{
```
Comment (d2lam) [inline launch_test.go]: isn't `screwdriver.Secrets` an array?

File: launch.go [hunk 4]
```diff
@@ -256,19 +255,59 @@ func launch(api screwdriver.API, buildID, rootDir, emitterPath string) error {
...
 	return nil
 }
 
+func createEnvironment(base map[string]string, secrets screwdriver.Secrets) []string {
+	combined := map[string]string{}
+
+	// Start with the current environment
+	for _, e := range os.Environ() {
+		s := strings.Split(e, "=")
+		if len(s) != 2 {
```
Comment (stjohnjohnson) [inline launch.go]:
```
$ export FOO="BAR=BAN"
$ env | grep FOO 
FOO=BAR=BAN
```

File: launch.go [hunk 5]
```diff
@@ -256,19 +255,59 @@ func launch(api screwdriver.API, buildID, rootDir, emitterPath string) error {
...
+			log.Printf("WARN: bad environment value from base environment: %s", e)
```
Comment (stjohnjohnson) [inline launch.go]: So this is just a warning, it continues anyway.

File: screwdriver/screwdriver.go (hunk 1 above)
Comment (stjohnjohnson) [inline screwdriver/screwdriver.go]: it's done

Comment (d2lam) [inline screwdriver/screwdriver.go]: redeployed?

Comment (d2lam) [inline screwdriver/screwdriver.go]: You merged with 1 thumb tsk tsk @stjohnjohnson

**Comment by stjohnjohnson** [comment]
I'm not sold on the string splitting and the warning about the environment variables, but as a first pass I'm cool with 👍 

simpsons

File: launch.go (hunk 4 above)
Comment (d2lam) [inline launch.go]: Perhaps:

```
i := strings.Index(e, "=")
if ( i == -1 ) {// error }
k := e[:i]
v := e[i+1:]
combined[k] = v
```

**Comment by d2lam** [comment]
👍

File: launch_test.go (hunk 3 above)
Comment (jer) [inline launch_test.go]: It's backed by an array. It's a slice of screwdriver.Secret

**Pull request merged**